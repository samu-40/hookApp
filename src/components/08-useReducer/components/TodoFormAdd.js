import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from '../../../hooks/useForm';

const TodoFormAdd = ({handleAdd}) => {

    const [{description}, handleInputChanges, reset] = useForm({description:''});

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {return};

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };

        handleAdd(newTodo);

        reset();

    };


    return (
        <>
            <h6>Agregar TODO</h6>
            <hr />

            <form
                className='d-grid gap-2'
                onSubmit={handleSubmit}
            >
                <input
                    type='text'
                    name='description'
                    placeholder='TODO....'
                    autoComplete='off'
                    className='form-control'
                    value={description}
                    onChange={handleInputChanges}
                />

                <button
                    type='submit'
                    className='btn btn-outline-success mt-1 '
                >
                    Add
                        </button>
            </form>
        </>
    );
};

TodoFormAdd.propTypes = {
    handleAdd: PropTypes.func.isRequired,
}

export default TodoFormAdd
