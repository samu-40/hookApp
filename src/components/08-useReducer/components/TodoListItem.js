import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({todo, handleDelete, handleToggleDone, index}) => {

    return (
        <li
            className='list-group-item'
        >

            <p
                className={`${todo.done && 'complete'}`}
                onClick={() => handleToggleDone(todo.id)}
            >
                {index + 1}. {todo.desc}
            </p>

            <button
                className='btn btn-danger'
                onClick={() => handleDelete(todo.id)}
            >
                Delete
            </button>

        </li>
    );

};

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggleDone: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
};
