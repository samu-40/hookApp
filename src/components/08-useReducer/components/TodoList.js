import React from 'react'
import { TodoListItem } from './TodoListItem'
import PropTypes from 'prop-types';

export const TodoList = ({todos, handleDelete, handleToggleDone}) => {

    return (
        <ul>
            {
                todos.map((todo, i) => (
                    <TodoListItem 
                        key={todo.id}
                        todo={todo}
                        handleDelete={handleDelete}
                        handleToggleDone={handleToggleDone}
                        index={i}
                    />
                ))
            }
        </ul>
    );

};

TodoListItem.propTypes = {
    todos: PropTypes.array,
    handleDelete: PropTypes.func.isRequired,
    handleToggleDone: PropTypes.func.isRequired,
};
