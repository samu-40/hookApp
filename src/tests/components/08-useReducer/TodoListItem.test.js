import {shallow} from 'enzyme';
import {TodoListItem} from '../../../components/08-useReducer/components/TodoListItem'
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en TodoListItem.js', () => {
    const handleDelete = jest.fn();
    const handleToggleDone = jest.fn();
    const index = 0

    const wrapper = shallow(
        <TodoListItem
            todo={demoTodos[0]}
            handleDelete={handleDelete}
            handleToggleDone={handleToggleDone}
            index={index}
        />
    );
        
    const id = demoTodos[0].id;
    const btn = wrapper.find('.btn');
    const p = wrapper.find('p');

    test('Debe mostrarse correctamente el componente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe llamase handleDelete', () => {
        btn.simulate('click');

        expect(handleDelete).toHaveBeenCalledWith(id);
    });

    test('Debe llamase handleToggleDone', () => {
        p.simulate('click');

        expect(handleToggleDone).toHaveBeenCalledWith(id);
    });

    test('Mostrar el texto correctamente', () => {
        const text = p.text().trim();

        expect(text).toBe(`${index+1}. ${demoTodos[0].desc}`);
    });    

    test('Debe tener la clase complete si TODO.done = true', () => {
        const todo = demoTodos[0];
        todo.done = true;
        
        const wrapper = shallow(
            <TodoListItem
                todo={todo}
                handleDelete={()=>{}}
                handleToggleDone={()=>{}}
                index={0}

            />
        );
        
        const clase = wrapper.find('p').hasClass('complete');

        expect(clase).toBe(true);
    })
    

});
