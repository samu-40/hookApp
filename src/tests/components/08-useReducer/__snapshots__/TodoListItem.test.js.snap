// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoListItem.js Debe mostrarse correctamente el componente 1`] = `
<li
  className="list-group-item"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Aprender pintura
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
